management/imapimport.py

```diff
@@ -0,0 +1,136 @@
+#!/usr/local/lib/mailinabox/env/bin/python
+# Gmail/IMAP import/export
+
+import os
+import sys
+import re
+import shutil
+import subprocess
+import string
+import random
+
+
+def run_cmd(config, info):
+	from decimal import Decimal, getcontext
+	getcontext().prec = 3
+
+	args = ['su', 'mail', '-g', 'mail', '-s', '/bin/bash', '-c']
+	exe = f'/usr/local/lib/mailinabox/env/bin/offlineimap -c {config} '
+	if info:
+		exe += '--info'
+	else:
+		exe += '-o'
+
+	args.append(exe)
+	print('')
+	output = ''
+
+	# example: Copy message UID 118742 (8/137) Source:[Gmail]/All Mail -> In:.Import
+	regex = re.compile(r'Copy message UID.*?\((\d+)/(\d+)\)')
+
+	try:
+		process = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+		for line in iter(process.stdout.readline, b''):
+			text = line.decode('utf8')
+
+			# output progress %
+			match = regex.search(text)
+			if match:
+				percent = ( Decimal(match.group(1)) / Decimal(match.group(2)) ) * 100
+				print(f'\r{percent}% ', end='')
+			else:
+				output += text
+
+			if process.returncode:
+				break
+	except Exception:
+		process.kill()
+		raise
+
+	print('\rDone!\n')
+	return output
+
+def importexport_cmdline(exec, cmd, type, user, remoteuser, remotepass, remotehost, remoteport, *folderlist, **options):
+	try:
+		remoteport = int(remoteport)
+		user, sep, domain = user.rpartition('@')
+		if cmd not in ('import', 'export') or type not in ('Gmail', 'IMAP') or not user:
```

Comment:
It's important to expect that users will not always use properly captialized terms. For example, `gmail` instead of `Gmail` and `imap` instead of `IMAP`. We should account and correct for these.

```suggestion
		if cmd.lower() not in ('import', 'export') or type.lower() not in ('gmail', 'imap') or not user:
```

Furthermore consider creating more specific error messages for each of these failure conditions (`cmd`, `type` and `user`)

Reply:
I don't think this will work because the parameters are used to render the config file: render = text.format(**context)

Realistically, I don't really expect people to use the commandline version. It really needs a UI where all this can be drop-downs and validated fields

Reply:
> Realistically, I don't really expect people to use the commandline version.

Until a TUI or GUI is implemented people will use the command line because that's the only way they can do so.

So I'd rather have a good CLI right off the bat regardless if we're going to also make this available within roundcube or the admin panel.

Reply:
I've made the errors a bit more specific.

Adding too much intelligence like automatically completing the gmail imap server is going to make it more difficult to implement a UI because the parameters will move around.

I don't have time to implement any more changes, sorry.